src/main/java/org/apache/commons/codec/binary/Base16.java

```diff
@@ -0,0 +1,251 @@
+        if (availableChars == 1 && availableChars == dataLen) {
+            context.ibitWorkArea = decodeOctet(data[offset]) + 1;   // store 1/2 byte for next invocation of decode, we offset by +1 as empty-value is 0
+            return;
+        }
+
+        // we must have an even number of chars to decode
+        final int charsToProcess = availableChars % BYTES_PER_ENCODED_BLOCK == 0 ? availableChars : availableChars - 1;
+
+        final byte[] buffer = ensureBufferSize(charsToProcess / BYTES_PER_ENCODED_BLOCK, context);
+
+        int result;
+        int i = 0;
+        if (dataLen < availableChars) {
+            // we have 1/2 byte from previous invocation to decode
+            result = (context.ibitWorkArea - 1) << BITS_PER_ENCODED_BYTE;
+            result |= decodeOctet(data[offset++]);
+            i = 2;
+
+            buffer[context.pos++] = (byte)result;
+
+            // reset to empty-value for next invocation!
+            context.ibitWorkArea = 0;
+        }
+
+        while (i < charsToProcess) {
+            result = decodeOctet(data[offset++]) << BITS_PER_ENCODED_BYTE;
+            result |= decodeOctet(data[offset++]);
+            i += 2;
+            buffer[context.pos++] = (byte)result;
+        }
+
+        // we have one char of a hex-pair left over
+        if (i < dataLen) {
+            context.ibitWorkArea = decodeOctet(data[i]) + 1;   // store 1/2 byte for next invocation of decode, we offset by +1 as empty-value is 0
+        }
+    }
+
+    private int decodeOctet(final byte octet) {
+        int decoded = -1;
+        if (octet >= 0 && octet < decodeTable.length) {
+            decoded = decodeTable[octet];
+        }
+
+        if (decoded == -1) {
+            throw new IllegalArgumentException("Invalid octet in encoded value: " + (int)octet);
+        }
+
+        return decoded;
+    }
+
+    @Override
+    void encode(final byte[] data, final int offset, final int length, final Context context) {
+        if (context.eof) {
+            return;
+        }
+
+        if (length < 0) {
+            context.eof = true;
+            return;
+        }
+
+        final byte[] buffer = ensureBufferSize(length * BYTES_PER_ENCODED_BLOCK, context);
+
+        final int end = offset + length;
+        for (int i = offset; i < end; i++) {
+            final int value = data[i];
+            final int high = (value >> BITS_PER_ENCODED_BYTE) & MASK_4BITS;
+            final int low = value & MASK_4BITS;
+            buffer[context.pos++] = encodeTable[high];
+            buffer[context.pos++] = encodeTable[low];
+        }
+    }
+
+    /**
+     * Returns whether or not the {@code octet} is in the Base16 alphabet.
+     *
+     * @param octet The value to test.
+     *
+     * @return {@code true} if the value is defined in the the Base16 alphabet {@code false} otherwise.
+     */
+    @Override
+    public boolean isInAlphabet(final byte octet) {
```

Comment:
Yes exactly! I want this to be spec compliant, not something that might produce unexpected behaviour. If you don't know if the hex you are decoding is upper or lower case, I think you have bigger problems... i.e. are you sure it is even valid hex?

Reply:
I think there are two issues here:

- Not knowing if the hex is upper or lower case, but it must use exclusively one or the other
- Not allowing mixed case hex

Ideally the most flexible decoder would have a 3rd mode for case insensitive decoding. This would apply to decoding only. It would work using a 3rd decoding table that combines the upper and lower case tables.

However allowing mixed case in the same input is not really valid. So the decoder would have to detect the first A-F character and ensure the same case is used for the rest. An exception can be raised if not. This can even be covered by the strict/lenient decoding policy. Tracking the first A-F character is a lot of extra code so I don't like this idea.

What do you think to using the decoding policy to cover this? The constructor accepts the upper/lower case flag for encoding and the strict/lenient decoding flag. If using strict decoding the case must match. If lenient mode then decoding is case insensitive. After all, if we are allowing a trailing character to be discarded, then allowing upper or lower case hex seems no worse a violation of a true encoding.

Reply:
I think it is not a bad suggestion, but I also think this is perhaps feature/scope creep. My goal was to implement the decoder according to the RFC. What you have suggested sounds like an additional feature which could be added later without issue. I would rather get this PR merged as it is. Is that okay with you?

Reply:
OK.

The idea of using the lenient decoding to cover both case insensitive decoding and allowing trailing characters is a mash up. A case insensitive decoding table can be added as an option later. After all, you may wish to support case insensitive decoding but enforce no trailing characters.